apis/stacks/v1alpha1/stackdefinition.go

```diff
@@ -0,0 +1,120 @@
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StackDefinitionSpec defines the desired state of StackDefinition
+type StackDefinitionSpec struct {
+	AppMetadataSpec `json:",inline"`
+	CRDs            CRDList                 `json:"customresourcedefinitions,omitempty"`
+	Controller      ControllerSpec          `json:"controller,omitempty"`
+	Permissions     PermissionsSpec         `json:"permissions,omitempty"`
+	Behavior        StackDefinitionBehavior `json:"behavior,omitempty"`
+}
+
+// StackDefinitionSource is the stack image which this stack configuration is from.
+// In the future, other source types may be supported, such as a URL.
+type StackDefinitionSource struct {
+	// a container image id
+	Image string `json:"image,omitempty"`
+}
+
+// StackDefinitionBehavior specifies the behavior for the stack, if the stack has behaviors instead of a controller
+type StackDefinitionBehavior struct {
+	CRD    StackDefinitionBehaviorCRD       `json:"crd,omitempty"`
+	Engine StackResourceEngineConfiguration `json:"engine,omitempty"`
+	// Theoretically, source and engine could be specified at a per-hook level
+	// as well.
+	Source    StackDefinitionSource            `json:"source,omitempty"`
+	Reconcile StackDefinitionHookConfiguration `json:"reconcile,omitempty"`
+}
+
+// StackDefinitionBehaviorCRD represents the CRD which the stack's behavior controller will watch. When CRs of this
+// CRD kind appear and are modified in the cluster, the behavior will execute.
+type StackDefinitionBehaviorCRD struct {
+	APIVersion string `json:"apiVersion"`
+	Kind       string `json:"kind"`
+}
+
+// StackResourceEngineConfiguration represents a configuration for a resource engine, such as helm2 or kustomize.
+type StackResourceEngineConfiguration struct {
+	// Type is the engine type, such as "helm2" or "kustomize"
+	Type string `json:"type"`
+	// Because different engine configurations could specify conflicting field names,
+	// we want to namespace the engines with engine-specific keys
+	Kustomize *KustomizeEngineConfiguration `json:"kustomize,omitempty"`
+}
+
+// KustomizeEngineConfiguration provides kustomize-specific engine configuration.
+type KustomizeEngineConfiguration struct {
+	Overlays      []ResourceEngineOverlay    `json:"overlays,omitempty"`
+	Kustomization *unstructured.Unstructured `json:"kustomization,omitempty"`
+}
+
+// ResourceEngineOverlay configures the stack behavior controller to transform the input CR into some output objects
+// for the underlying resource engine. This is expected to be interpreted by the engine-specific logic in the controller.
+type ResourceEngineOverlay struct {
+	APIVersion string                         `json:"apiVersion"`
```

Comment:
> User needs to be able to specify the namespace, too

Do you have a concrete example? I was trying to think of one but I just have some contrived ones, so I'm not sure whether they're real.

> I think inlining metav1.ObjectReference makes it easy

Hmmmmm good idea - I do like the idea of reusing something, but I'm not sure about including some of the other fields in `ObjectReference`, like `UID` and `FieldPath`. I would lean toward not using it so that those fields aren't added in this case.

Reply:
> Do you have a concrete example?

I think it'd be useful for apps that deploy to multiple namespaces. What happens is that they include `Namespace` resources in the app bundle and refer to it in different resources. So, you might get two resources with same name but different namespaces. Though I am not sure how common this pattern is. I usually just go for the safer option, in this case just making it `ObjectReference`, which is something we already use for namespaced references in the codebase, but both approaches are just fine.